<a href="https://colab.research.google.com/github/Paulmora23/ENTREGABLE_BIT_PAULAMOLINA/blob/main/Entregable_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


##LIMPIEZA BASE DE DATOS ANIMAL DATA DIRTY


Archivo esperado:/content/animal_data_dirty1.csv

In [ ]:
import pandas as pd

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving animal_data_dirty1 (1).csv to animal_data_dirty1 (1).csv


In [ ]:
import pandas as pd
df = pd.read_csv('animal_data_dirty1 (1).csv', delimiter=';')

df.head()

,Animal type,Country,Weight kg,Body Length cm,Gender,Animal code,Latitude,Longitude,Animal name,Observation date,Data compiled by
0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,03.01.2024,James Johnson
1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,03.02.2024,James Johnson
2,European bison,Poland,930.0,335.0,male,NaN,52.828845,23.820144,Szefu,01.03.2024,Anne Anthony
3,European bison,Poland,909.0,311.0,not determined,NaN,52.830509,23.826849,NaN,01.03.2024,Anne Anthony
4,European bison™,Poland,581.0,277.0,female,NaN,52.834109,23.807093,NaN,01.03.2024,Anne Anthony


In [ ]:
df.shape

(1011, 11)

#1-✅DIAGNÓSTICO INICIAL

#Cuántas filas y columnas tiene el DataFrame?

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1011 entries, 0 to 1010
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Animal type       991 non-null    object 
 1   Country           999 non-null    object 
 2   Weight kg         984 non-null    float64
 3   Body Length cm    984 non-null    float64
 4   Gender            992 non-null    object 
 5   Animal code       0 non-null      float64
 6   Latitude          913 non-null    float64
 7   Longitude         913 non-null    float64
 8   Animal name       52 non-null     object 
 9   Observation date  1011 non-null   object 
 10  Data compiled by  1011 non-null   object 
dtypes: float64(5), object(6)
memory usage: 87.0+ KB


##Cuántas filas y columnas tiene el DataFrame?
- El Dataframe tiene 1011 filas y 11 columnas

##Cuáles son las columnas y qué tipo de dato tiene cada una?

 Las columnas son 11 y tiene los siguientes tipos de datos.

✔Animal type:Tipo de dato es objeto en formato texto y contiene valores no   nulos.

✔Country:Tipo de dato es objeto en formato texto y contiene valores no nulos.

**✔Weight Kg:** Tipo de dato flotante  en numero decimal y contiene valores no nulos.

✔Body Length cm: **texto en negrita** Tipo de dato flotante en numero decimal y contiene valores no nulos.

**✔Gender:** Tipo de dato es objeto en formato texto y contiene valores no nulos.

**✔Animal code:** Tipo de dato flotante en numero decimal y contiene valores no nulos (columna que esta totalmente vacia por lo que no tiene ningun registro).

**✔Latitude:** Tipo de dato de dato flotante en numero decimal y contiene valores no nulos.

**✔Longitude:** Tipo de dato flotante en numero decimal y contiene valores no nulos.

✔Animal name:Tipo de dato es objeto en formato texto y contiene valores no nulos.

✔Observation date:Tipo de dato es objeto en formato texto y contiene formato fecha y contiene valores no nulos.

✔Data compiled by: Tipo de dato es objeto en formato texto y contiene valores no nulos.


In [ ]:
df.describe()

,Weight kg,Body Length cm,Animal code,Latitude,Longitude
count,984.000000,984.000000,0.0,913.000000,913.000000
mean,39.745503,39.107724,NaN,49.393369,18.203280
std,156.290076,58.628601,NaN,7.168900,3.899601
min,-0.252000,-19.000000,NaN,-78.582973,11.074008
25%,0.293000,19.000000,NaN,48.186913,14.384559
50%,0.331500,21.000000,NaN,49.560723,18.944015
75%,0.800000,23.000000,NaN,52.212433,21.033243
max,1100.000000,350.000000,NaN,52.853843,34.896734


#Cuántos valores faltantes existen por columna?

In [ ]:
df.isnull().sum()

,0
Animal type,20
Country,12
Weight kg,27
Body Length cm,27
Gender,19
Animal code,1011
Latitude,98
Longitude,98
Animal name,959
Observation date,0


- Segun la tabla de valores muestra por columna los datos faltantes de cada uno, del cual el dato donde mas faltan datos es en Animal code con 1011 registros y los que no tienen son Observation date y Data compiled by.

#Cuántos registros duplicados existen por columna?

In [ ]:
df.duplicated().sum()

np.int64(167)

Existen 167 registros duplicados por columna

#Escribe una breve explicación de al menos 3 problemas de calidad que hayas encontrado.

In [ ]:
# 1. Explorar inconsistencias en la especie del animal
display(df["Animal type"].value_counts(dropna=False))
# Alternativa: df["Animal type"].unique()

# 2. Explorar inconsistencias en el país
display(df["Country"].value_counts(dropna=False))
# Alternativa: df["Country"].unique()

# 3. Explorar inconsistencias en el género
display(df["Gender"].value_counts(dropna=False))

# 4. Explorar quién registró los datos
display(df["Data compiled by"].value_counts(dropna=False))

,count
Animal type,
red squirrel,579
hedgehog,278
lynx,71
European bison,63
NaN,20


,count
Country,
Poland,291
Germany,176
Slovakia,151
Hungary,146
Czech Republic,103
Austria,74
PL,15
NaN,12
HU,11


,count
Gender,
male,499
female,488
NaN,19
not determined,5


,count
Data compiled by,
Anne Anthony,473
Bob Bobson,259
John Johnson,249
James Johnson,30


#De los valores analisados se encuentran los siguientes:

- Se encuentran inconsistencias categóricas  y de digitación por ejemplo en Country esta "Poland" y el acronimo de PL; generando información incorrecta puesto que tecnicamente nuestro codigo lo va a generar como dos variables diferentes y En Animal type hay errores tipograficos.

- Se encuentran valores nulos como en la categoria Animal Code donde el 100% de los registros estan vacios.

- Se encuentran en las primeras dos filas iniciales registros fantasmas  sin información y variables como Weight Kg y Body Lenght cm presentan ausencias que distorcionan un analisis real de la situación.

- Existen 167 filas exactamente repetidas que sesgan un conteo real de los datos observados.

- La categoria observacion se evidencia los datos relacionados como numero general y la mitad de la otra base con formato fecha lo que ocasiona distorcion al momento de la base ser manipulada.

#2-**✅LIMPIEZA DE DATOS**

In [ ]:
df[df["Animal name"].isnull()].head(10)

,Animal type,Country,Weight kg,Body Length cm,Gender,Animal code,Latitude,Longitude,Animal name,Observation date,Data compiled by
0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,03.01.2024,James Johnson
1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,03.02.2024,James Johnson
3,European bison,Poland,909.000,311.0,not determined,NaN,52.830509,23.826849,NaN,01.03.2024,Anne Anthony
4,European bison,Poland,581.000,277.0,female,NaN,52.834109,23.807093,NaN,01.03.2024,Anne Anthony
5,European bison,Poland,900.000,295.0,male,NaN,52.834759,23.817201,NaN,01.03.2024,Anne Anthony
6,European bison,Poland,620.000,250.0,female,NaN,52.834960,23.820750,NaN,01.03.2024,Anne Anthony
7,lynx,Hungary,24.000,76.0,male,NaN,48.505494,20.551966,NaN,01.03.2024,Anne Anthony
8,lynx,Hungary,23.000,82.0,female,NaN,48.506529,20.554548,NaN,01.03.2024,Anne Anthony
10,red squirrel,PL,0.348,20.0,female,NaN,52.242060,21.057802,NaN,01.03.2024,Anne Anthony
12,red squirrel,Poland,0.340,21.0,NaN,NaN,52.240284,21.051948,NaN,01.03.2024,Anne Anthony


Se determina cambiar los valores nulos en la columa de Animal name por "Desconocida".

In [ ]:
df["Animal name"]= df["Animal name"].fillna("Desconocida")

Se determina eliminar la columna Animal code puesto que no tienen ningun dato del cual nos pueda afectar nuestra base de datos y nos permite simplificar la información.

In [ ]:
df = df.drop(columns=['Animal code'])
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 991 entries, 2 to 1010
Data columns (total 10 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Animal type       991 non-null    object 
 1   Country           991 non-null    object 
 2   Weight kg         976 non-null    float64
 3   Body Length cm    976 non-null    float64
 4   Gender            984 non-null    object 
 5   Latitude          913 non-null    float64
 6   Longitude         913 non-null    float64
 7   Animal name       991 non-null    object 
 8   Observation date  991 non-null    object 
 9   Data compiled by  991 non-null    object 
dtypes: float64(4), object(6)
memory usage: 85.2+ KB


Se decide eliminar de Animal Type y de Country las filas que estan nulas; puesto que no infiere ningun dato.

In [ ]:
df = df.dropna(subset=["Animal type"])

In [ ]:
df = df.dropna(subset=["Country"])

In [ ]:
df.isnull().sum()

,0
Animal type,0
Country,0
Weight kg,15
Body Length cm,15
Gender,7
Latitude,78
Longitude,78
Animal name,0
Observation date,0
Data compiled by,0


Despues de hacer la limpieza de datos de los valores nulos en nuestra base se evidencio que se generan 166 duplicados que se ejecutaran de la siguiente forma:

In [ ]:
df.duplicated().sum()

np.int64(166)

In [ ]:
df = df.drop_duplicates()

df.shape

(825, 10)

In [ ]:
df[df.duplicated()].head(10)

,Animal type,Country,Weight kg,Body Length cm,Gender,Latitude,Longitude,Animal name,Observation date,Data compiled by


#Se determina hayar la mediana de la categoria Weight Kg y Body Lenght cm

In [ ]:
mediana_weight = df["Weight kg"].median()
df["Weight kg"] = df["Weight kg"].fillna(mediana_weight)


mediana_length = df["Body Length cm"].median()
df["Body Length cm"] = df["Body Length cm"].fillna(mediana_length)

#Estandarizar nombres en 'Animal type'

In [ ]:
dict_animales = {
    'red squirrell': 'red squirrel',
    'red squirel': 'red squirrel',
    'lynx?': 'lynx',
    'European bison™': 'European bison',
    'European bisson': 'European bison',
    'European buster': 'European bison',
    'ledgehod': 'hedgehog',
    'wedgehod': 'hedgehog'
    }

#Estandarizar nombres en 'Country'

In [ ]:
dict_paises = {
    'PL': 'Poland',
    'HU': 'Hungary',
    'Hungry': 'Hungary',
    'CZ': 'Czech Republic',
    'Czech': 'Czech Republic',
    'DE': 'Germany'
}

#Corregir valores negativos en medidas numéricas (errores de signos)

In [ ]:
df['Weight kg'] = df['Weight kg'].abs()
df['Body Length cm'] = df['Body Length cm'].abs()

In [ ]:
df['Weight kg'] = df.groupby('Animal type')['Weight kg'].transform(lambda s: s.fillna(s.median()))
df['Body Length cm'] = df.groupby('Animal type')['Body Length cm'].transform(lambda s: s.fillna(s.median()))

#Imputar nulos en variables de text

In [ ]:
df['Gender'] = df['Gender'].fillna('not determined')
df['Animal name'] = df['Animal name'].fillna('Desconocida')

#Convertir tipos de datos (Fechas)

In [ ]:
df_raw = pd.read_csv('animal_data_dirty1 (1).csv', sep=';', encoding='utf-8-sig')
df['Observation date'] = df_raw.loc[df.index, 'Observation date']

In [ ]:
df['Observation date'] = pd.to_datetime(df['Observation date'], format='mixed', dayfirst=True)

#Demostración de Transformaciones Correctas

In [ ]:
print("Forma final del DataFrame limpio:", df.shape)
print("\nDuplicados restantes:", df.duplicated().sum())

Forma final del DataFrame limpio: (825, 10)

Duplicados restantes: 0


In [ ]:
print("\nValores nulos restantes:")
print(df.isnull().sum())


Valores nulos restantes:
Animal type          0
Country              0
Weight kg            0
Body Length cm       0
Gender               0
Latitude            73
Longitude           73
Animal name          0
Observation date    32
Data compiled by     0
dtype: int64


In [ ]:
print("\nCategorías limpias en 'Animal type':")
print(df['Animal type'].value_counts())


Categorías limpias en 'Animal type':
Animal type
red squirrel      413
hedgehog          278
lynx               71
European bison     63
Name: count, dtype: int64


In [ ]:
print("\nCategorías limpias en 'Country':")
print(df['Country'].value_counts())


Categorías limpias en 'Country':
Country
Poland            192
Germany           178
Slovakia          151
Hungary           118
Czech Republic    112
Austria            66
Australia           4
CC                  4
Name: count, dtype: int64


In [ ]:
print("\nResumen estadístico de variables numéricas:")
display(df[['Weight kg', 'Body Length cm']].describe())


Resumen estadístico de variables numéricas:


,Weight kg,Body Length cm
count,825.000000,825.000000
mean,47.352070,42.860606
std,169.650992,63.258801
min,0.200000,11.000000
25%,0.298000,19.000000
50%,0.360000,21.000000
75%,1.000000,23.000000
max,1100.000000,350.000000


In [ ]:
df.isnull().sum()

,0
Animal type,0
Country,0
Weight kg,0
Body Length cm,0
Gender,0
Latitude,73
Longitude,73
Animal name,0
Observation date,0
Data compiled by,0


Tras aplicar las estrategias de limpieza e imputación, se evidencia que las variables cuentan con 0 valores nulos. En cuanto a latitud y longitud, se conservaron sus 73 valores ausentes sin imputar, ya que calcular una media o mediana en coordenadas  generaría ubicaciones inexactas, preservando así la integridad de los datos.

#✅3-Numpy

In [ ]:
import numpy as np

# Clasificación biológica según la masa corporal: 'Fauna Mayor' vs 'Fauna Menor'
df['Size_category'] = np.where(df['Weight kg'] >= 5.0, 'Fauna Mayor', 'Fauna Menor')

# Verificación de la distribución de la nueva variable
print("Distribución de la categoría de tamaño:")
display(df['Size_category'].value_counts())
display(df[['Animal type', 'Weight kg', 'Size_category']].head(10))

Distribución de la categoría de tamaño:


,count
Size_category,
Fauna Menor,691
Fauna Mayor,134


,Animal type,Weight kg,Size_category
2,European bison,930.000,Fauna Mayor
3,European bison,909.000,Fauna Mayor
4,European bison,581.000,Fauna Mayor
5,European bison,900.000,Fauna Mayor
6,European bison,620.000,Fauna Mayor
7,lynx,24.000,Fauna Mayor
8,lynx,23.000,Fauna Mayor
9,red squirrel,0.308,Fauna Menor
10,red squirrel,0.348,Fauna Menor
11,red squirrel,0.316,Fauna Menor


In [ ]:
# Cálculo de estadísticos descriptivos con NumPy sobre la variable Weight kg
mediana_peso = np.median(df['Weight kg'])
media_peso = np.mean(df['Weight kg'])
desv_peso = np.std(df['Weight kg'])

print(f"Mediana del peso general: {mediana_peso:.3f} kg")
print(f"Media (promedio) del peso general: {media_peso:.3f} kg")
print(f"Desviación estándar del peso general: {desv_peso:.3f} kg")

Mediana del peso general: 0.360 kg
Media (promedio) del peso general: 47.352 kg
Desviación estándar del peso general: 169.548 kg


Con np.where() se creó la variable Size_category, clasificando a los animales en Fauna Mayor ($\ge$ 5 kg; 134 registros) y Fauna Menor (< 5 kg; 691 registros).Al analizar Weight kg con NumPy, la mediana fue de 0.360 kg frente a un promedio de 47.352 kg (desviación estándar: 169.548 kg). Esta diferencia ocurre porque la gran mayoría de la muestra son ardillas y erizos de menos de 1 kg, mientras que el peso extremo de los bisontes distorsiona la media hacia arriba; por ello, la mediana representa mejor el valor central de la base de datos.

#4-✅Análisis con groupby

In [ ]:
# Conteo de registros por especie
obs_por_animal = df.groupby('Animal type').size().reset_index(name='Total observaciones')
obs_por_animal = obs_por_animal.sort_values(by='Total observaciones', ascending=False)
display(obs_por_animal)

,Animal type,Total observaciones
3,red squirrel,413
1,hedgehog,278
2,lynx,71
0,European bison,63


In [ ]:
# Promedio de peso (kg) y longitud corporal (cm) por especie
prom_por_animal = df.groupby('Animal type')[['Weight kg', 'Body Length cm']].mean().round(2).reset_index()
display(prom_por_animal)

,Animal type,Weight kg,Body Length cm
0,European bison,588.40,244.24
1,hedgehog,0.80,20.21
2,lynx,23.27,82.90
3,red squirrel,0.30,20.51


In [ ]:
# Promedio de peso y longitud según la ubicación geográfica
prom_por_pais = df.groupby('Country')[['Weight kg', 'Body Length cm']].mean().round(2).reset_index()
prom_por_pais = prom_por_pais.sort_values(by='Weight kg', ascending=False)
display(prom_por_pais)

,Country,Weight kg,Body Length cm
2,CC,634.00,272.75
6,Poland,130.00,74.00
3,Czech Republic,54.51,44.09
7,Slovakia,32.45,43.39
5,Hungary,2.88,26.47
1,Austria,2.14,25.08
0,Australia,0.49,20.25
4,Germany,0.46,20.85


In [ ]:
# Análisis adicional: Comparación entre machos y hembras por especie
analisis_genero = df.groupby(['Animal type', 'Gender'])[['Weight kg', 'Body Length cm']].agg(['count', 'mean']).round(2)
display(analisis_genero)

Weight kg         Body Length cm        
                                  count    mean          count    mean
Animal type    Gender                                                 
European bison female                35  481.11             35  223.11
               male                  26  728.00             26  273.65
               not determined         2  651.00              2  231.50
hedgehog       female               136    0.75            136   19.40
               male                 140    0.84            140   20.92
               not determined         2    0.95              2   25.00
lynx           female                37   21.00             37   82.35
               male                  34   25.74             34   83.50
red squirrel   female               191    0.30            191   20.40
               male                 217    0.30            217   20.61
               not determined         5    0.31              5   20.40

Este análisis multidimensional permite identificar dimorfismo sexual evidente en la fauna mayor: los bisontes machos promedian 728.00 kg frente a 481.11 kg de las hembras, mientras que en las ardillas rojas el peso medio entre machos y hembras es prácticamente idéntico (0.30 kg).

#✅5-INTERPRETACION DE RESULTADOS

**Concentración del muestreo en fauna menor**

**1- Qué se encontró**: La muestra está fuertemente dominada por especies pequeñas.

- Evidencia:Las ardillas rojas (red squirrel) con 413 registros (50.1%) y los erizos (hedgehog) con 278 registros (33.7%) suman 691 observaciones, representando el 83.8% de los datos limpios. En contraste, los linces (lynx, 71) y bisontes (European bison, 63) apenas suman el 16.2%.

**Asimetría y representatividad de las medidas estadísticas**

**2-Qué se encontró:** El peso promedio global no refleja la realidad de la mayoría de los individuos analizados.

- Evidencia: La mediana global es de 0.360 kg, mientras que el promedio general es de 47.352 kg con una desviación estándar de 169.548 kg. Además, el bisonte europeo promedia 588.40 kg (con máximos de 1.100 kg), frente a los 0.30 kg de la ardilla roja.

**Sesgo geográfico en el registro de biomasa**

**3-Qué se encontró:** Ciertos países exhiben promedios de peso desproporcionadamente elevados en comparación con los demás.

- Evidencia: Polonia registra un peso medio de 130.00 kg (192 registros) y República Checa 54.51 kg (112 registros), mientras que Alemania reporta una media de solo 0.46 kg (178 registros) y Hungría 2.88 kg (118 registros).

**Dimorfismo sexual evidente en grandes mamíferos vs. micromamíferos**

**4-Qué se encontró:** La diferencia de masa corporal entre machos y hembras varía drásticamente según la escala corporal del animal.

- Evidencia: En el bisonte europeo, los machos promedian 728.00 kg frente a los 481.11 kg de las hembras (una brecha de casi 247 kg, ~51% más pesados). Por el contrario, en las ardillas rojas el promedio para ambos sexos es idéntico: 0.30 kg (machos: 0.297 kg vs. hembras: 0.297 kg).